circuit_knitting/cutting/cut_finding/search_space_generator.py

```diff
@@ -0,0 +1,208 @@
+    space when stratified beam search is employed.  The stratum_func can be
+    None, in which case each level of the search has only one stratum, which
+    is then labeled None.
+
+    greedy_bound_func (lambda current_best_cost, *args) can be either
+    None or a function that computes upper bounds to costs that are used during
+    the greedy depth-first phases of search.  If None is provided, the upper
+    bound is taken to be infinity.  In greedy search, the search proceeds in a
+    greedy best-first, depth-first fashion until either a goal state is reached,
+    a deadend is reached, or the cost bound provided by the greedy_bound_func is
+    exceeded.  In the latter two cases, the search backjumps to the lowest cost
+    state in the search frontier and the search proceeds from there.  The
+    inputs passed to the greedy_bound_func are the current lowest cost in the
+    search frontier and the input arguments that were passed to the
+    optimizationPass() method of the search algorithm.  If the greedy_bound_func
+    simply returns current_best_cost, then the search behavior is equivalent to
+    pure best-first search.  Returning None is equivalent to returning an
+    infinite greedy bound, which produces a purely greedy best-first,
+    depth-first search.
+
+    next_state_func (lambda state, *args) is a function that returns a list
+    of next states generated from the input state.  An ActionNames object
+    should be incorporated into the additional input arguments in order to
+    generate next-states.
+
+    goal_state_func (lambda state, *args) is a function that returns True if
+    the input state is a solution state of the search.
+
+    upperbound_cost_func (lambda goal_state, *args) can either be None or a
+    function that returns an upper bound to the optimal cost given a goal_state
+    as input.  The upper bound is used to prune next-states from the search in
+    subsequent calls to the optimizationPass() method of the search algorithm.
+    If upperbound_cost_func is None, the cost of the goal_state as determined
+    by cost_func is used as an upper bound to the optimal cost.  If the
+    upperbound_cost_func returns None, the effect is equivalent to returning
+    an infinite upper bound (i.e., no cost pruning is performed on subsequent
+    calls to the optimizationPass method.
+
+    mincost_bound_func (lambda *args) can either be None or a function that
+    returns a cost bound that is compared to the minimum cost across all
+    vertices in a search frontier.  If the minimum cost exceeds the min-cost
+    bound, the search is terminated even if a goal state has not yet been found.
+    Returning None is equivalent to returning an infinite min-cost bound (i.e.,
+    min-cost checking is effectively not performed).  A mincost_bound_func that
+    is None is likewise equivalent to an infinite min-cost bound.
+    """
+
+    cost_func:  Callable  = None,
```

Comment:
We can make an attempt to fill out these Callable types. For example, ``Callable[[ArgAType, ArgBType], ReturnType]``

Reply:
Yes, it's on my list of things to do.

Reply:
For this particular module, commit 551237d87113102c741ae1aa1d84a4c519472da9 now has these Callable types filled in.